src/rhash/hash.c

```diff
@@ -2388,21 +2383,185 @@ static int rc_hash_gamecube(char hash[33], const rc_hash_iterator_t* iterator)
   {
     if (dol_sizes[ix] == 0)
       continue;
-
+    rc_file_seek(iterator, file_handle, part_offset + dol_offsets[ix], SEEK_SET);
     if (ix < 7)
       rc_hash_iterator_verbose_formatted(iterator, "Hashing %u byte main.dol code segment %u", dol_sizes[ix], ix);
     else
       rc_hash_iterator_verbose_formatted(iterator, "Hashing %u byte main.dol data segment %u", dol_sizes[ix], ix - 7);
+    remaining_size = dol_sizes[ix];
+    while (remaining_size > READ_CHUNK_SIZE)
+    {
+      rc_file_read(iterator, file_handle, buffer, READ_CHUNK_SIZE);
+      md5_append(md5, buffer, READ_CHUNK_SIZE);
+      remaining_size -= READ_CHUNK_SIZE;
+    }
+    rc_file_read(iterator, file_handle, buffer, (int32_t)remaining_size);
+    md5_append(md5, buffer, (int32_t)remaining_size);
+  }
+
+  free(buffer);
+  return 1;
+}
+
+static int rc_hash_nintendo_disc(char hash[33], const rc_hash_iterator_t* iterator)
+{
+  md5_state_t md5;
+  void* file_handle;
+
+  const uint32_t MAIN_HEADER_SIZE = 0x80;
+  const uint64_t REGION_CODE_ADDRESS = 0x4E000;
+  const uint32_t CLUSTER_SIZE = 0x7C00;
+  const uint32_t MAX_CLUSTER_COUNT = 1024 / 32;
+
+  uint32_t partition_info_table[8];
+  uint8_t total_partition_count = 0;
+  uint32_t* partition_table;
+  uint32_t tmd_offset;
+  uint32_t tmd_size;
+  uint32_t part_offset;
+  uint32_t part_size;
+
+  uint8_t quad_buffer[4];
+  uint8_t* buffer;
+
+  uint32_t ix, jx, kx;
+  uint8_t encrypted;
+
+  file_handle = rc_file_open(iterator, iterator->path);
+  if (!file_handle)
+    return rc_hash_iterator_error(iterator, "Could not open file");
+
+  /* Check Gamecube */
+  rc_file_seek(iterator, file_handle, 0x1c, SEEK_SET);
+  rc_file_read(iterator, file_handle, quad_buffer, 4);
+  if (quad_buffer[0] == 0xC2 && quad_buffer[1] == 0x33 && quad_buffer[2] == 0x9F && quad_buffer[3] == 0x3D)
+  {
+    md5_init(&md5);
+    if (rc_hash_nintendo_disc_partition(&md5, iterator, file_handle, 0, 0) == 0)
+      return 0;
+
+    rc_file_close(iterator, file_handle);
 
-    rc_file_seek(iterator, file_handle, dol_offsets[ix], SEEK_SET);
-    rc_file_read(iterator, file_handle, buffer, dol_sizes[ix]);
+    return rc_hash_finalize(iterator, &md5, hash);
+  }
 
-    md5_append(&md5, buffer, dol_sizes[ix]);
+  /* Check Wii */
+  rc_file_seek(iterator, file_handle, 0x18, SEEK_SET);
+  rc_file_read(iterator, file_handle, quad_buffer, 4);
+  if (quad_buffer[0] != 0x5D || quad_buffer[1] != 0x1C || quad_buffer[2] != 0x9E || quad_buffer[3] != 0xA3)
+  {
+    rc_file_close(iterator, file_handle);
+    return rc_hash_iterator_error(iterator, "Not a supported Nintendo disc");
+  }
+
+  /* Check encryption byte - if 0x61 is 0, disc is encrypted */
+  rc_file_seek(iterator, file_handle, 0x61, SEEK_SET);
+  rc_file_read(iterator, file_handle, quad_buffer, 1);
+  encrypted = (quad_buffer[0] == 0);
+
+  /* Hash main headers */
+  buffer = (uint8_t*)malloc(MAIN_HEADER_SIZE);
```

Comment:
See related comment about allocating a single CLUSTER_SIZE buffer for the length of this function.